File: packages/modules/chess/go/core/board/board.go
package board

import (
	"strings"

	"github.com/hieudoanm/chess/core/types"
	"github.com/hieudoanm/chess/core/utils"
)

func EmptyBoard() types.Board {
	return types.Board{}
}

func CloneBoard(b types.Board) types.Board {
	return b
}

func PutPiece(b types.Board, piece *types.Piece, sq types.Square) {
	b[sq] = piece
}

func GetPiece(b types.Board, sq types.Square) *types.Piece {
	return b[sq]
}

func RemovePiece(b types.Board, sq types.Square) {
	b[sq] = nil
}

func FindKing(b types.Board, color types.Color) (types.Square, bool) {
	for sq := 0; sq < 64; sq++ {
		p := b[sq]
		if p != nil && p.Color == color && p.Type == types.King {
			return sq, true
		}
	}
	return 0, false
}

func pieceFromChar(ch byte) *types.Piece {
	var pieceType types.PieceType
	switch {
	case ch == 'p' || ch == 'P':
		pieceType = types.Pawn
	case ch == 'n' || ch == 'N':
		pieceType = types.Knight
	case ch == 'b' || ch == 'B':
		pieceType = types.Bishop
	case ch == 'r' || ch == 'R':
		pieceType = types.Rook
	case ch == 'q' || ch == 'Q':
		pieceType = types.Queen
	case ch == 'k' || ch == 'K':
		pieceType = types.King
	default:
		return nil
	}
	color := types.Black
	if ch < 'a' {
		color = types.White
	}
	return &types.Piece{Color: color, Type: pieceType}
}

func BoardFromFen(boardPart string) types.Board {
	b := EmptyBoard()
	rows := strings.Split(boardPart, "/")
	for r := 0; r < 8; r++ {
		row := rows[r]
		f := 0
		for i := 0; i < len(row); i++ {
			ch := row[i]
			if ch >= '1' && ch <= '8' {
				f += int(ch - '0')
			} else {
				piece := pieceFromChar(ch)
				if piece != nil {
					b[utils.Square(7-r, f)] = piece
				}
				f++
			}
		}
	}
	return b
}

func BoardToFen(b types.Board) string {
	var sb strings.Builder
	for r := 7; r >= 0; r-- {
		empty := 0
		for f := 0; f < 8; f++ {
			piece := b[utils.Square(r, f)]
			if piece != nil {
				if empty > 0 {
					sb.WriteByte(byte('0' + empty))
					empty = 0
				}
				if piece.Color == types.White {
					sb.WriteByte(byte(piece.Type[0] - 32))
				} else {
					sb.WriteByte(byte(piece.Type[0]))
				}
			} else {
				empty++
			}
		}
		if empty > 0 {
			sb.WriteByte(byte('0' + empty))
		}
		if r > 0 {
			sb.WriteByte('/')
		}
	}
	return sb.String()
}

var PieceValues = map[types.PieceType]int{
	types.Pawn:   100,
	types.Knight: 320,
	types.Bishop: 330,
	types.Rook:   500,
	types.Queen:  900,
	types.King:   20000,
}

var PieceUnicode = map[string]string{
	"K": "\u2654",
	"Q": "\u2655",
	"R": "\u2656",
	"B": "\u2657",
	"N": "\u2658",
	"P": "\u2659",
	"k": "\u265A",
	"q": "\u265B",
	"r": "\u265C",
	"b": "\u265D",
	"n": "\u265E",
	"p": "\u265F",
}

File: packages/modules/chess/ts/src/board/board.ts
import type { Board, Piece, Square, PieceType, Color } from '../types/types';
import { toSquare } from '../utils/utils';

export const createEmptyBoard = (): Board => Array(64).fill(null);

export const cloneBoard = (board: Board): Board => [...board];

export const setPiece = (board: Board, piece: Piece, sq: Square): void => {
  board[sq] = piece;
};

export const getPiece = (board: Board, sq: Square): Piece | null =>
  board[sq] ?? null;

export const removePiece = (board: Board, sq: Square): void => {
  board[sq] = null;
};

export const findKing = (board: Board, color: Color): Square | null => {
  for (let sq = 0; sq < 64; sq++) {
    const piece = board[sq];
    if (piece && piece.color === color && piece.type === 'k') return sq;
  }
  return null;
};

const pieceFromChar = (ch: string): Piece | null => {
  const typeMap: Record<string, PieceType> = {
    p: 'p',
    n: 'n',
    b: 'b',
    r: 'r',
    q: 'q',
    k: 'k',
  };
  const lower = ch.toLowerCase();
  const type = typeMap[lower];
  if (!type) return null;
  return { color: ch === lower ? 'b' : 'w', type };
};

export const fromFenBoard = (boardPart: string): Board => {
  const board = createEmptyBoard();
  const rows = boardPart.split('/');
  for (let r = 0; r < 8; r++) {
    const row = rows[r]!;
    let f = 0;
    for (const ch of row) {
      if (ch >= '1' && ch <= '8') {
        f += parseInt(ch);
      } else {
        const piece = pieceFromChar(ch);
        if (piece) board[toSquare(7 - r, f)] = piece;
        f++;
      }
    }
  }
  return board;
};

export const toFenBoard = (board: Board): string => {
  let fen = '';
  for (let r = 7; r >= 0; r--) {
    let empty = 0;
    for (let f = 0; f < 8; f++) {
      const piece = board[toSquare(r, f)];
      if (piece) {
        if (empty > 0) {
          fen += empty;
          empty = 0;
        }
        fen += piece.color === 'w' ? piece.type.toUpperCase() : piece.type;
      } else {
        empty++;
      }
    }
    if (empty > 0) fen += empty;
    if (r > 0) fen += '/';
  }
  return fen;
};

export const PIECE_VALUES: Record<PieceType, number> = {
  p: 100,
  n: 320,
  b: 330,
  r: 500,
  q: 900,
  k: 20000,
};

export const PIECE_UNICODE: Record<string, string> = {
  K: '♔',
  Q: '♕',
  R: '♖',
  B: '♗',
  N: '♘',
  P: '♙',
  k: '♚',
  q: '♛',
  r: '♜',
  b: '♝',
  n: '♞',
  p: '♟',
};
